File: scripts/algorithms/Design Add and Search Words Data Structure/Design Add and Search Words Data Structure.py
class TrieNode:
    
    def __init__(self, val):
        self.val = val
        self.children = {}
        self.isEnd = False


class WordDictionary:

    def __init__(self):
        self.root = TrieNode("*")
        
    def addWord(self, word: str) -> None:
        
        curr = self.root
        
        for c in word:
            
            if c not in curr.children:
                curr.children[c] = TrieNode(c)
                
            curr = curr.children[c]
        
        curr.isEnd = True
        
    def search(self, word: str) -> bool:
        
        def dfs(root, word):
            curr = root

            for i in range(len(word)):

                if word[i] == ".":
                    for l in curr.children.values():
                        if dfs(l, word[i+1:]) == True:
                            return True
                        
                    return False

                if word[i] not in curr.children:
                    return False

                curr = curr.children[word[i]]

            return curr.isEnd
        
        return dfs(self.root, word)

File: scripts/algorithms/D/Design Add and Search Words Data Structure/Design Add and Search Words Data Structure.cpp
// Runtime: 2662 ms (Top 14.83%) | Memory: 623.6 MB (Top 10.10%)
class WordDictionary {
public:
    struct tr {
        bool isword;
        vector<tr*> next;
        tr() {
            next.resize(26, NULL);
            isword = false;
        }
    };

    tr *root;
    WordDictionary() {
        root = new tr();
    }

    void addWord(string word) {
        tr *cur = root;

        for(auto c: word) {
            if(cur->next[c-'a'] == NULL) {
                cur->next[c-'a'] = new tr();
            }
            cur = cur->next[c-'a'];
        }

        cur->isword = true;
    }

    bool dfs(string& word, int i, tr* cur) {
        if(!cur ){
            return false;
        }

        if(i == word.size()) {
            return cur->isword;
        }

        for (; i < word.size(); i++) {
            char c = word[i];
            if (c == '.') {
                for (auto it: cur->next) {
                    if(dfs(word, i+1, it)) {
                        return true;
                    }
                }
                return false;
            }
            else {
                if(cur->next[c-'a'] == NULL) {
                    return false;
                }
                cur = cur->next[c-'a'];
            }
        }

        return cur->isword;
    }

    bool search(string word) {
        return dfs(word, 0, root);
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */